From ac0af4f737a9e11832805d965be51963bc7026d9 Mon Sep 17 00:00:00 2001
From: Eric Sandeen <sandeen@redhat.com>
Date: Mon, 4 Aug 2014 19:04:41 -0400
Subject: [fs] btrfs: fix error handling in create_pending_snapshot

Message-id: <1407179109-13714-395-git-send-email-sandeen@redhat.com>
Patchwork-id: 87329
O-Subject: [RHEL7.1 PATCH 394/422] btrfs: fix error handling in create_pending_snapshot
Bugzilla: 1071179
RH-Acked-by: Zach Brown <zab@redhat.com>

commit 47a306a74842248dcd537b85f9a36c7b156c59a9
Author: Eric Sandeen <sandeen@redhat.com>
Date:   Thu Jun 12 00:53:44 2014 -0500

    btrfs: fix error handling in create_pending_snapshot

    fcebe456 cut and pasted some code to a later point
    in create_pending_snapshot(), but didn't switch
    to the appropriate error handling for this stage
    of the function.

    Signed-off-by: Eric Sandeen <sandeen@redhat.com>
    Signed-off-by: Chris Mason <clm@fb.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/btrfs/transaction.c b/fs/btrfs/transaction.c
index 1c54e2e..ac984a3 100644
--- a/fs/btrfs/transaction.c
+++ b/fs/btrfs/transaction.c
@@ -1284,11 +1284,13 @@ static noinline int create_pending_snapshot(struct btrfs_trans_handle *trans,
   goto fail;
  }
 
- pending->error = btrfs_qgroup_inherit(trans, fs_info,
-           root->root_key.objectid,
-           objectid, pending->inherit);
- if (pending->error)
-  goto no_free_objectid;
+ ret = btrfs_qgroup_inherit(trans, fs_info,
+       root->root_key.objectid,
+       objectid, pending->inherit);
+ if (ret) {
+  btrfs_abort_transaction(trans, root, ret);
+  goto fail;
+ }
 
  /* see comments in should_cow_block() */
  set_bit(BTRFS_ROOT_FORCE_COW, &root->state);
-- 
1.7.1